client/state/data-layer/wpcom/account-recovery/lookup/index.js

```diff
@@ -17,60 +12,44 @@ import {
 	updatePasswordResetUserData,
 } from 'state/account-recovery/reset/actions';
 
-export const fromApi = data => [
+/*
+ * Uses `substring()` to force string values. Anything
+ * other than a string will throw, thus invalidating the parse
+ */
+export const fromApi = ( { primary_email, primary_sms, secondary_email, secondary_sms } ) => [
 	{
-		email: data.primary_email,
-		sms: data.primary_sms,
+		email: primary_email.substring( 0 ),
```

Comment:
TIL

Reply:
me too! I was unsure about the pattern but it seems to hold strongly enough the same way we use `!!` for booleans or `|0` for integers.

Reply:
interesting. one big diff vs `~~` or `|0` is that this throws instead of coercing. Is that what you want? is a null primary email acceptable?

Reply:
> this throws instead of coercing

yes! actually I should have made that clearer in my comment. the previous code performed the same check except it did it through `isString()`

throwing here means we'll fail-over into `onError` - the exception won't raise beyond `dispatchRequestEx`

Reply:
Is throwing a runtime `TypeError`s really a good practice? I think it's something like causing a segfault -- there are better ways to handle errors. I feel that the `substring` trick is more suitable for an obfuscated code contest and would always prefer a traditional `if ( ! isString( ... ) ) throw Error( ... )`.

Reply:
@jsnajdr in a sense I think it's okay and also dirty. the fine part here though is that the `fromApi` system _wants_ us to throw errors in order to escape and _we do this all the time with arrays and objects_.

that is, if I expect the response to be an array I'm going to implicitly depend on it being so and will write `postIds: data.posts.map( ( { ID } ) => id )` - I wouldn't expect any of us to write this…

```js
if ( ! data || ! data.posts || ! Array.isArray( data.posts ) ) {
	throw new Error( 'invalid or missing `posts` field' );
}

const postIds = data.posts.map( post => {
	if ( ! post || ! post.hasOwnProperty( 'ID' ) ) {
		throw new Error( 'missing ID property on post item' );
	}

	return post.ID;
} );
```

that's an extreme case of course. I was simply trying to find a balance in this PR and I didn't think it necessitated the verbose throw.